src/types.cc

```diff
@@ -0,0 +1,27 @@
+#include "types.hpp"
+#include <iostream>
+
+std::ostream& operator<<(std::ostream& out, const DeviceMode value){
+    const char* s = 0;
+#define PROCESS_VAL(p) case(p): s = #p; break;
+    switch(value){
+        PROCESS_VAL(NO_DEVICE_MODE);
+        PROCESS_VAL(GPU_DEVICE_MODE);
+        PROCESS_VAL(CPU_DEVICE_MODE);
+    }
+#undef PROCESS_VAL
+    return out << s;
+}
+
```

Comment:
Wow nice construction :zap:  ...short but not the most intuitive way, need to think about it ^^